File: backend/core/logger.py
"""
Global Logger for StrikeIQ Realtime Trading System
Structured logging with service prefixes and trace IDs
"""

import logging
import uuid
from datetime import datetime
from typing import Optional

# Global trace context
class TraceContext:
    def __init__(self):
        self.trace_id: Optional[str] = None
    
    def start_trace(self) -> str:
        self.trace_id = str(uuid.uuid4())[:8].upper()
        return self.trace_id
    
    def get_trace_id(self) -> str:
        return self.trace_id or "NO_TRACE"
    
    def clear_trace(self):
        self.trace_id = None

# Global trace context
trace_context = TraceContext()

# Custom formatter with trace ID
class TraceFormatter(logging.Formatter):
    def format(self, record):
        # Safely get optional attributes
        service = getattr(record, "service", "APP")
        trace = getattr(record, "trace_prefix", "")
        
        # Add trace ID if available
        trace_id = trace_context.get_trace_id()
        if trace_id != "NO_TRACE":
            trace = f"[TRACE {trace_id}] "
        else:
            trace = ""
        
        # Format with service prefix
        return f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] [{record.levelname}] [{service}] {trace}{record.getMessage()}"

# Configure logging
def setup_logger(service_name: str) -> logging.Logger:
    """Setup logger with service name and trace formatting"""
    logger = logging.getLogger(service_name)
    
    if not logger.handlers:
        # Create handler
        handler = logging.StreamHandler()
        handler.setFormatter(TraceFormatter())
        
        # Add handler to logger
        logger.addHandler(handler)
        logger.setLevel(logging.DEBUG)
        
        # Prevent duplicate logs
        logger.propagate = False
    
    # Create adapter with service context
    return logging.LoggerAdapter(logger, {'service': service_name})

# Service-specific loggers
upstox_logger = setup_logger("UPSTOX")
protobuf_logger = setup_logger("PROTOBUF")
aggregator_logger = setup_logger("AGGREGATOR")
ws_logger = setup_logger("WS")
auth_logger = setup_logger("AUTH")
market_logger = setup_logger("MARKET")
api_logger = setup_logger("API")

# Helper functions for trace management
def start_trace() -> str:
    """Start a new trace and return trace ID"""
    return trace_context.start_trace()

def get_trace_id() -> str:
    """Get current trace ID"""
    return trace_context.get_trace_id()

def clear_trace():
    """Clear current trace"""
    trace_context.clear_trace()

# Decorator for automatic trace management
def with_trace(func):
    """Decorator to automatically manage trace for a function"""
    def wrapper(*args, **kwargs):
        trace_id = start_trace()
        try:
            result = func(*args, **kwargs)
            return result
        finally:
            clear_trace()
    return wrapper

# Export all loggers and utilities
__all__ = [
    'upstox_logger',
    'protobuf_logger', 
    'aggregator_logger',
    'ws_logger',
    'auth_logger',
    'market_logger',
    'api_logger',
    'start_trace',
    'get_trace_id',
    'clear_trace',
    'with_trace'
]

File: frontend/src/utils/logger.ts
/**
 * Frontend Logger for StrikeIQ Realtime Trading System
 * Structured logging with service prefixes and trace tracking
 */

export interface LogContext {
  traceId?: string;
  component?: string;
  [key: string]: any;
}

class Logger {
  private static instance: Logger;
  private currentTraceId: string | null = null;

  static getInstance(): Logger {
    if (!Logger.instance) {
      Logger.instance = new Logger();
    }
    return Logger.instance;
  }

  startTrace(): string {
    this.currentTraceId = Math.random().toString(36).substring(2, 10).toUpperCase();
    return this.currentTraceId;
  }

  getTraceId(): string {
    return this.currentTraceId || 'NO_TRACE';
  }

  clearTrace(): void {
    this.currentTraceId = null;
  }

  private formatMessage(level: string, service: string, message: string, context?: LogContext): string {
    const timestamp = new Date().toISOString();
    const traceId = context?.traceId || this.getTraceId();
    const tracePrefix = traceId !== 'NO_TRACE' ? `[TRACE ${traceId}] ` : '';
    
    // Format context as key=value pairs
    const contextStr = context ? 
      Object.entries(context)
        .filter(([key]) => key !== 'traceId')
        .map(([key, value]) => `${key}=${value}`)
        .join(' ') : '';
    
    const contextSuffix = contextStr ? ` ${contextStr}` : '';
    
    return `[${timestamp}] [${level}] [${service}] ${tracePrefix}${message}${contextSuffix}`;
  }

  debug(service: string, message: string, context?: LogContext): void {
    console.debug(this.formatMessage('DEBUG', service, message, context));
  }

  info(service: string, message: string, context?: LogContext): void {
    console.info(this.formatMessage('INFO', service, message, context));
  }

  warn(service: string, message: string, context?: LogContext): void {
    console.warn(this.formatMessage('WARN', service, message, context));
  }

  error(service: string, message: string, context?: LogContext): void {
    console.error(this.formatMessage('ERROR', service, message, context));
  }
}

// Global logger instance
const logger = Logger.getInstance();

// Service-specific loggers
export const wsLogger = {
  debug: (message: string, context?: LogContext) => logger.debug('WS', message, context),
  info: (message: string, context?: LogContext) => logger.info('WS', message, context),
  warn: (message: string, context?: LogContext) => logger.warn('WS', message, context),
  error: (message: string, context?: LogContext) => logger.error('WS', message, context),
};

export const storeLogger = {
  debug: (message: string, context?: LogContext) => logger.debug('STORE', message, context),
  info: (message: string, context?: LogContext) => logger.info('STORE', message, context),
  warn: (message: string, context?: LogContext) => logger.warn('STORE', message, context),
  error: (message: string, context?: LogContext) => logger.error('STORE', message, context),
};

export const apiLogger = {
  debug: (message: string, context?: LogContext) => logger.debug('API', message, context),
  info: (message: string, context?: LogContext) => logger.info('API', message, context),
  warn: (message: string, context?: LogContext) => logger.warn('API', message, context),
  error: (message: string, context?: LogContext) => logger.error('API', message, context),
};

export const uiLogger = {
  debug: (message: string, context?: LogContext) => logger.debug('UI', message, context),
  info: (message: string, context?: LogContext) => logger.info('UI', message, context),
  warn: (message: string, context?: LogContext) => logger.warn('UI', message, context),
  error: (message: string, context?: LogContext) => logger.error('UI', message, context),
};

// Trace management
export const traceManager = {
  startTrace: () => logger.startTrace(),
  getTraceId: () => logger.getTraceId(),
  clearTrace: () => logger.clearTrace(),
};

// Higher-order function for automatic trace management
export function withTrace<T extends any[], R>(
  fn: (...args: T) => R,
  serviceName?: string
): (...args: T) => R {
  return (...args: T): R => {
    const traceId = logger.startTrace();
    
    if (serviceName) {
      logger.info(serviceName, 'TRACE STARTED', { traceId });
    }
    
    try {
      const result = fn(...args);
      
      if (serviceName) {
        logger.info(serviceName, 'TRACE COMPLETED', { traceId });
      }
      
      return result;
    } catch (error) {
      if (serviceName) {
        logger.error(serviceName, 'TRACE ERROR', { traceId, error: String(error) });
      }
      throw error;
    } finally {
      logger.clearTrace();
    }
  };
}

export default logger;
